### DISCUSSION TITLE
DMA: Memory-to-Memory

### DISCUSSION THREAD

**Discussion by chaosAD** [joncampbell123/dosbox-x#4500] [Q&A]
Hi there, I tried DMA memory-to-memory transfer using channels 0 and 1 on the 1st DMA. But it does not seem to work. The channel count registers remain unchanged after software DMA request for channel 0 is issued. Is this feature emulated as well? Or is there some setting need to be done?

**Comment by chaosAD** [comment]
Ah! No wonder it doesn't work. The source code has a `TODO` remark for the memory-to-memory (under the `REQuest register` 0x9) in the _dma.cpp_ file. Also the `Command register` 0x8 in the `WriteControllerReg(.)` was left unimplemented, which is important for configuring the memory-to-memory transfer.

**Comment by joncampbell123** [comment]
> Hi there, I tried DMA memory-to-memory transfer using channels 0 and 1 on the 1st DMA. But it does not seem to work. The channel count registers remain unchanged after software DMA request for channel 0 is issued. Is this feature emulated as well? Or is there some setting need to be done?

You're the first to try it :)

Memory to memory DMA is so seldom used in PC software that most emulators, DOSBox included, don't bother with it. Past the mid 1990s there may be motherboard chipsets that don't implement it either (such chipsets I recall had also dropped support for FM encoded floppy support thus making them unusable with 8-inch floppy drives, therefore making them MFM-only floppy controllers).

The reason it isn't used is because, even on original IBM hardware, it was *slower* than just using the CPU to copy memory, so there wasn't any point to using it.

**Comment by joncampbell123** [comment]
On a related note, it is also unusual to use the direction bit to run DMA "backwards" through memory, but that is implemented because there is one 1991 demoscene entry that actually does render Sound Blaster DMA output backwards in memory like that :)

EDIT: This demo: http://files.scene.org/get/mirrors/hornet/demos/1992/emf_id.zip, and Pouet entry here https://www.pouet.net/prod.php?which=6171.

Notice that in any other DOSBox branch, if you can get it to do Sound Blaster output at all, it comes out as static and noise because SVN doesn't pay attention to the direction bit. Except Sound Blaster output may not work because the demo *also* relies on the DSP alias port 22Dh rather than 22Ch to command the DSP chip (Sound Blaster Pro and lower cards decode only bits 3-1 when responding to ports 22xh except for some ports like the mixer chip and therefore the same I/O ports appear on even and odd numbers).
